nullaway/src/main/java/com/uber/nullaway/handlers/CompositeHandler.java

```diff
@@ -184,7 +184,10 @@ public NullnessHint onDataflowVisitMethodInvocation(
       NullnessHint n =
           h.onDataflowVisitMethodInvocation(
               node, state, apContext, inputs, thenUpdates, elseUpdates, bothUpdates);
-      nullnessHint = nullnessHint.merge(n);
+      nullnessHint = nullnessHint.mergeByPriority(n);
+      if (nullnessHint.isHighestPriority()) {
+        break;
```

Comment:
I don't know if this will currently break anything, but this is **not** a safe optimization.

 It means that calls to `.onDataflowVisitMethodInvocation(...)` might be skipped when a handler early in the chain "wins" in terms of determining the `NullnessHint`. But keep in mind that handler methods are not meant to be pure, they are meant to be callbacks for events of the analysis. In fact, `.onDataflowVisitMethodInvocation(...)` specifically is often called for its effects on its `AccessPathNullnessPropagation.Updates` paramenters.

For example, [AssertionHandler](https://github.com/uber/NullAway/blob/ed0ef72cdf0a166993ba2703c0d63f41eff97f22/nullaway/src/main/java/com/uber/nullaway/handlers/AssertionHandler.java#L46) will _always_ return `NullnessHint.UNKNOWN`, because it doesn't actually care about the nullability of the statement `assertThat(foo).bar()`, but will instead add whatever facts that assertion statement indicates about `foo` into `bothUpdates`.

Reply:
good point - thanks the explanation, not sure how i missed this :sweat: 

so for `assertThat(x).isNotNull()` when the handlers examine the `isNotNull()` method invocation part, we dont really care about the nullness hint the handlers return but we want to give them the ability to examine the whole method-call-chain and realize that `x` is non-null afterwards (i.e. `bothUpdates`)

Reply:
Almost correct. It's not about the method call chain, but the [chain of handlers](https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/handlers/Handlers.java#L44). The way these handlers are used is, whenever you call a method like `.isNotNull()`, each handler in turn has the opportunity to process it. Some handlers, like `LibraryModelsHandler` will process it by i.e. looking up some information and deciding whether or not the return of method call should `@Nullable` or `@NonNull`. But that's not the only action to be taken on such call. As you know, `AssertionHandler` will instead look at the receiver of `.isNotNull()` and unpack `x` within a `assertThat(x)` receiver and then set the nullability of `x` (within dataflow after the current statement).

With your change, if a handler early in the chain computes a `FORCE_NONNULL` nullability for the method call's return, then no handler after that has a chance to process the call and perhaps set other state or any of the updates passed as arguments based on the call.